pkg/rancher-components/src/components/StringList/StringList.vue

```diff
@@ -0,0 +1,573 @@
+
+        return;
+      }
+      if (this.isEditItem) {
+        this.toggleEditMode(false);
+
+        return;
+      }
+      if (this.selected) {
+        const index = findStringIndex(this.items, this.selected, false);
+
+        if (index !== -1) {
+          /**
+           * Select the next item in the list when an item is to be deleted.
+           */
+          const item = (this.items[index + 1] || this.items[index - 1]);
+
+          this.onSelect(item);
+          this.setFocus(item);
+
+          this.deleteItem(this.items[index]);
+        }
+      }
+    },
+
+    setFocus(refId: string) {
+      this.$nextTick(() => this.getElemByRef(refId)?.focus());
+    },
+
+    /**
+     * Move scrollbar when the selected item is over the top or bottom side of the box
+     */
+    moveScrollbar(arrow: Arrow, value?: number) {
+      const box = this.getElemByRef(BOX);
+      const item = this.getElemByRef(this.selected || '');
+
+      if (box && item && item.className.includes(CLASS.item)) {
+        const boxRect = box.getClientRects()[0];
+        const itemRect = item.getClientRects()[0];
+
+        if (
+          (arrow === 'down' && itemRect.bottom > boxRect.bottom) ||
+          (arrow === 'up' && itemRect.top < boxRect.top)
+        ) {
+          const top = (value ?? itemRect.height) * DIRECTION[arrow];
+
+          box.scrollBy({ top });
+        }
+      }
+    },
+
+    /**
+     * Set an error and eventually assign an error class to the element
+     */
+    toggleError(type: Error, val: boolean, refId?: string) {
+      this.errors[type] = val;
+      if (refId) {
+        this.toggleErrorClass(refId, val);
+      }
+    },
+
+    toggleErrorClass(refId: string, val: boolean) {
+      const input = this.getElemByRef(refId)?.$el;
+
+      if (input) {
+        if (val) {
+          input.classList.add(CLASS.error);
+        } else {
+          input.classList.remove(CLASS.error);
+        }
+      }
+    },
+
+    /**
+     * Show/Hide the input line to create new item
+     */
+    toggleCreateMode(show: boolean) {
+      if (show) {
+        this.value = '';
+
+        this.isCreateItem = true;
+        this.setFocus(INPUT.create);
+      } else {
+        this.value = null;
+        this.toggleError('duplicate', false);
+        this.onSelectLeave();
+
+        this.isCreateItem = false;
+      }
+    },
+
+    /**
+     * Show/Hide the in-line editing to edit an existing item
+     */
+    toggleEditMode(show: boolean, item?: string) {
+      if (show) {
+        this.toggleCreateMode(false);
+        this.value = this.isEditItem;
+
+        this.isEditItem = item || '';
+        this.setFocus(INPUT.edit);
+      } else {
+        this.value = null;
+        this.toggleError('duplicate', false);
+        this.onSelectLeave();
+
+        this.isEditItem = null;
+      }
+    },
+
+    getElemByRef(id: string) {
+      const ref = this.$refs[id];
+
+      return (Array.isArray(ref) ? ref[0] : ref) as any;
+    },
```

Comment:
I'm noticing two different usages of `getElemByRef`:

1. Line 238: `const input = this.getElemByRef(refId)?.$el`
2. Lines 209-210:
   ``` 
      const box = this.getElemByRef(BOX);
      const item = this.getElemByRef(this.selected || '');
   ```

I think we can improve the type of `getElemByRef` can be improved by introducing a new function (or changing the return type of the existing function by always returning `this.$refs[id].$el`). We could probably attempt something like this to make sure that we are working with a stronger type than `any` 

```suggestion
    getElemByRef(id: string): Vue {
      const ref = this.$refs[id];

      return Array.isArray(ref) ? (ref[0] as Vue) : ref as Vue;
    },

    getHtmlDivElement(id: string) {
      return this.getElemByRef(id).$el as HTMLDivElement;
    },
```

Reply:
I missed that, good point